internal/pkg/coordinator/monitor.go

```diff
@@ -394,3 +426,72 @@ func runCoordinatorOutput(ctx context.Context, cord Coordinator, bulker bulk.Bul
 		}
 	}
 }
+
+func runUnenroller(ctx context.Context, bulker bulk.Bulk, policyId string, unenrollTimeout time.Duration, l zerolog.Logger, checkInterval time.Duration, agentsIndex string) {
+	t := time.NewTimer(checkInterval)
+	defer t.Stop()
+	for {
+		select {
+		case <-t.C:
+			if err := runUnenrollerWork(ctx, bulker, policyId, unenrollTimeout, l, agentsIndex); err != nil {
+				l.Err(err).Dur("unenroll_timeout", unenrollTimeout).Msg("failed to unenroll offline agents")
+			}
+			t.Reset(checkInterval)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func runUnenrollerWork(ctx context.Context, bulker bulk.Bulk, policyId string, unenrollTimeout time.Duration, l zerolog.Logger, agentsIndex string) error {
+	agents, err := dl.FindOfflineAgents(ctx, bulker, policyId, unenrollTimeout, dl.WithIndexName(agentsIndex))
+	if err != nil {
+		return err
+	}
+	agentIds := make([]string, len(agents))
+	for i, agent := range agents {
+		err = unenrollAgent(ctx, bulker, &agent, agentsIndex)
+		if err != nil {
+			return err
+		}
+		agentIds[i] = agent.Id
+	}
+	if len(agentIds) > 0 {
+		l.Info().Strs("agents", agentIds).Msg("marked agents unenrolled due to unenroll timeout")
+	}
+	return nil
+}
+
+func unenrollAgent(ctx context.Context, bulker bulk.Bulk, agent *model.Agent, agentsIndex string) error {
+	now := time.Now().UTC().Format(time.RFC3339)
+	fields := bulk.UpdateFields{
+		dl.FieldActive:           false,
+		dl.FieldUnenrolledAt:     now,
+		dl.FieldUnenrolledReason: unenrolledReasonTimeout,
+		dl.FieldUpdatedAt:        now,
+	}
+	body, err := fields.Marshal()
+	if err != nil {
+		return err
+	}
+	apiKeys := getAPIKeyIDs(agent)
+	if len(apiKeys) > 0 {
+		err = apikey.Invalidate(ctx, bulker.Client(), apiKeys...)
+		if err != nil {
```

Comment:
does it need to stop invalidation and unenrollment if it fails to invalidate one of the api keys?

Reply:
I think this is the correct behavior, because if it fails to invalidate then it will stop. The loop will run again after the check interval and then hopefully that time it will succeed. If it keeps failing then it will not unenroll Elastic Agent's which will then be obvious on the listing and that would prompt them to check on what is going wrong.